### PR TITLE
Update makefile for proto related rules

### PR DESCRIPTION
**Pull request by richardwxn**
1. Clean up the makefile for proto related rules, add README for patch instruction
2. Fetch the k8s protos dependencies to /tmp for it to work with container. Also update the proto path to include common image's shared protos
3. replace gofast plugin with gogofast plugin, which helps to fix the enum field unmarshalling issue

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -36,6 +36,7 @@ GOBIN ?= /work/out/bin
 RUN = docker run -t --sig-proxy=true -u $(UID) --rm \
 	-v /etc/passwd:/etc/passwd:ro \
 	-v $(readlink /etc/localtime):/etc/localtime:ro \
+	-v $(GOPATH)/src/istio.io/operator:/go/src/istio.io/operator \
```
Comment (sdake) [inline Makefile]: this will be overwritten when `make common` is run to update the makefiles, which may be often.

File: Makefile.core.mk [hunk 2]
```diff
@@ -78,37 +78,29 @@ clean-proto:
...
 
-# get imported protos to $GOPATH
 get_dep_proto:
-	GO111MODULE=off go get k8s.io/api/core/v1 k8s.io/api/autoscaling/v2beta1 k8s.io/apimachinery/pkg/apis/meta/v1/ github.com/gogo/protobuf/...
+	GO111MODULE=off GOPATH=/tmp go get k8s.io/api/core/v1 k8s.io/api/autoscaling/v2beta1 k8s.io/apimachinery/pkg/apis/meta/v1/ github.com/gogo/protobuf/...
 
-proto_iscp:
-	protoc -I=${GOPATH}/src -I./pkg/apis/istio/v1alpha2/ --proto_path=pkg/apis/istio/v1alpha2/ --gofast_out=pkg/apis/istio/v1alpha2/ pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto
+proto_iscp: get_dep_proto
+	protoc -I=/tmp/src -I=${GOPATH}/src/ -I=/usr/include/protobuf --gogofast_out=${GOPATH}/src ${GOPATH}/src/istio.io/operator/pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto
```
Comment (sdake) [inline Makefile.core.mk]: the first part looks good.  the ${GOPATH}/src - understand what your after, I 'll need to think a little bit about how we can generalize this.  I just woke up so I am not super sharp atm, but I think we need a system host bindmount on either `${GOPATH}` or `/`.`${GOPATH}`may work for our use case with operator repo, but may not work for every use case.  THe bindmount should be `rshared` - so will need to test if that will work as I expect.

I guess we can start with that, and expand.

Cheers
-steve

File: Makefile.core.mk [hunk 3]
```diff
@@ -78,37 +78,29 @@ clean-proto:
...
 	sed -i -e 's|github.com/gogo/protobuf/protobuf/google/protobuf|github.com/gogo/protobuf/types|g' pkg/apis/istio/v1alpha2/istiocontrolplane_types.pb.go
 	patch pkg/apis/istio/v1alpha2/istiocontrolplane_types.pb.go < pkg/apis/istio/v1alpha2/fixup_go_structs.patch
 
-proto_iscp_orig:
-	protoc -I=${GOPATH}/src -I./pkg/apis/istio/v1alpha2/ --proto_path=pkg/apis/istio/v1alpha2/ --gofast_out=pkg/apis/istio/v1alpha2/ pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto
+proto_iscp_orig: get_dep_proto
```
Comment (sdake) [inline Makefile.core.mk]: do we need this variant, or can we just delete it?

File: Makefile.core.mk [hunk 4]
```diff
@@ -78,37 +78,29 @@ clean-proto:
 
 # is all of this needed or is this cruft?
 
-# get imported protos to $GOPATH
 get_dep_proto:
-	GO111MODULE=off go get k8s.io/api/core/v1 k8s.io/api/autoscaling/v2beta1 k8s.io/apimachinery/pkg/apis/meta/v1/ github.com/gogo/protobuf/...
+	GO111MODULE=off GOPATH=/tmp go get k8s.io/api/core/v1 k8s.io/api/autoscaling/v2beta1 k8s.io/apimachinery/pkg/apis/meta/v1/ github.com/gogo/protobuf/...
```
Comment (sdake) [inline Makefile.core.mk]: interesting solutoin, although these deps are pulled automatically by for example `make download`or are already likely in the image .  Protobuf itself is in the expected systemwide location.

File: Makefile.core.mk (hunk 3 above)
Comment (ostromart) [reply Makefile.core.mk]: i think we should get rid of it. for cases where the patch file needs to be updated we can leave a README in the proto dir on how to update it. this just clutters up the makefile and is confusing to the average user.

File: Makefile.core.mk [hunk 5]
```diff
@@ -78,37 +78,29 @@ clean-proto:
...
 
-proto_iscporig_setup: get_dep_proto proto_iscp_orig
-
-proto_iscp_setup: get_dep_proto proto_iscp
-
-proto_values_setup: get_dep_proto proto_values
-
-proto_valuesorig_setup: get_dep_proto proto_values_orig
 
 gen_patch_iscp:
```
Comment (ostromart) [inline Makefile.core.mk]: same here - could we get rid of the gen_patch commands and move the process to a README in the proto dir?

File: Makefile.core.mk (hunk 3 above)
Comment (richardwxn) [reply Makefile.core.mk]: ok I will go remove it and add README

File: Makefile.core.mk (hunk 4 above)
Comment (richardwxn) [reply Makefile.core.mk]: The k8s deps are not available in the common image now. Actually pulling in those deps to /tmp using go get does not require changes to existing proto import paths.

File: Makefile.core.mk (hunk 5 above)
Comment (sdake) [reply Makefile.core.mk]: agreed any simplification of the `Makefile.core.mk` file is beneficial.  If things remain, they should be commented to indicate their purpose.

File: Makefile (hunk 1 above)
Comment (sdake) [reply Makefile]: I edited the above comment to indicate the correct make target that results in an overwrite of the Makefile.

File: Makefile.core.mk (hunk 4 above)
Comment (sdake) [reply Makefile.core.mk]: The common image does include parts of `gogo/protobuf`: https://github.com/istio/tools/blob/master/docker/build-tools/Dockerfile#L46-L49

if parts of `gogo/protbuf` are not present, perhaps you could improve thecommon-tools image to include what is needed?  We really want to pin any dependencies needed using either go.mod or build-tools.

The dependencies are in `go.mod` referenced as `k8s.io/api` and `k8s.io/apimachinery`.  Do you mean these are not pulled into the go package cache when using the repository?  Or just that using `-I` is not desireable with `protoc`?

File: Makefile (hunk 1 above)
Comment (richardwxn) [reply Makefile]: ok I will move this logic to core instead

File: Makefile.core.mk (hunk 4 above)
Comment (richardwxn) [reply Makefile.core.mk]: the protoc include path does not handle go package cache properly, that's why I have to pull those deps to temp directory

**Comment by richardwxn** [comment]
@sdake Can you help please take another look?

File: Makefile [hunk 6]
```diff
@@ -71,7 +71,7 @@ else
 export GOBIN ?= ./out/bin
 endif
 
-MAKE = $(RUN) make -e -f Makefile.core.mk
+MAKE = $(RUN) make --no-print-directory -e -f Makefile.core.mk
```
Comment (sdake) [inline Makefile]: This looks a little redundant: https://github.com/istio/common-files/commit/4328839147bab7f10dd5790bd95c275482de83d2

File: pkg/apis/istio/v1alpha2/README.md [hunk 7]
```diff
@@ -0,0 +1,18 @@
+### proto file patch instruction
+Currently there are two proto files defined temporarily in istio/operator: istiocontrolplane_types.proto and values.types.proto, which would be moved to istio/api after 1.3
+
+There are some type of fields in the original proto files, e.g. []map[string]interface{}, cannot be handled properly by protoc,
+so we comment out this fields and patch them back in the generated pb.go files
```
Comment (sdake) [inline pkg/apis/istio/v1alpha2/README.md]: zug...

File: Makefile (hunk 6 above)
Comment (richardwxn) [reply Makefile]: removed

**Comment by richardwxn** [comment]
> Thanks for the work!
> 
> I'm not sure this PR is quite right, but it removes a bunch of cruft, so approving - and we can iterate towards a better solution that works with CONTAINER_BUILD=1.
> 
> Cheers
> -steve

Thanks for checking. Now it works with both container and local.

**Pull request merged**